src/plugins/backend/backend.c

```diff
@@ -0,0 +1,949 @@
+
+
+#include "backend.h"
+#include "backendprivate.h"
+
+#include <kdberrors.h>
+#include <kdbprivate.h>
+
+int setMountpoint (BackendHandle * bh, KeySet * config, Key * errorKey)
+{
+	Key * root;
+
+	ksRewind (config);
+
+	root = ksNext (config);
+
+	Key * searchMountpoint = keyDup (root);
+	keyAddBaseName (searchMountpoint, "mountpoint");
+	Key * foundMountpoint = ksLookup (config, searchMountpoint, 0);
+
+	keyDel (searchMountpoint);
+
+	if (!foundMountpoint)
+	{
+		ELEKTRA_ADD_PLUGIN_MISBEHAVIOR_WARNINGF (errorKey, "Could not find mountpoint within root %s", keyName (root));
+		return -1;
+	}
+
+	bh->mountpoint = keyNew ("", KEY_VALUE, keyBaseName (root), KEY_END);
+	elektraKeySetName (bh->mountpoint, keyString (foundMountpoint), KEY_CASCADING_NAME | KEY_EMPTY_NAME);
+
+	keySetName (errorKey, keyName (bh->mountpoint));
+
+	if (!bh->mountpoint)
+	{
+		ELEKTRA_ADD_PLUGIN_MISBEHAVIOR_WARNINGF (errorKey, "Could not create mountpoint with name '%s' and value %s",
+							 keyString (foundMountpoint), keyBaseName (root));
+		return -1;
+	}
+
+	keyIncRef (bh->mountpoint);
+	return 0;
+}
+
+KeySet * processConfig (BackendHandle * bh, KeySet * config, Key * errorKey)
+{
+	if (setMountpoint (bh, config, errorKey) == -1)
+	{
+		return NULL;
+	}
+
+	KeySet * systemConfig = ksRenameKeys (config, "system");
+	ksDel (config);
+
+	return systemConfig;
+}
+
+int linkedListPosition (Key * cur, Key * errorKey)
+{
+	char * name = elektraMalloc (sizeof (keyBaseName (cur)));
```

Comment:
You are using `sizeof` wrong (not just here, but in many places). In C `sizeof` is evaluated by the compiler, so it cannot evaluate to the length of a string. Instead `sizeof (keyBaseName (cur)) == sizeof (char *)`, i.e. the size of a pointer (4 or 8 bytes depending on 32-bit or 64-bit machine).

In general you should use `strlen (myString) + 1` instead of `sizeof (myString)`, but when dealing with keys there is often a better solution, size `Key`s store there sizes. For example there are `keyGetBaseNameSize`, `keyGetNameSize` and `keyGetValueSize` (they all count the terminating zero-byte, so are the same as strlen + 1).

Reply:
Thank you for helping out.